docs: add NetworkPolicy samples page (#2085)
kyverno/website: src/content/docs/docs/guides/security.md

:::

The samples assume Kyverno is installed in the `kyverno` namespace. Each controller uses these ports:

| Controller              | Webhook | Metrics |
| ----------------------- | ------- | ------- |
| `admission-controller`  | 9443    | 8000    |
| `cleanup-controller`    | 9443    | 8000    |
| `background-controller` | —       | 8000    |
| `reports-controller`    | —       | 8000    |

Egress requirements are shared across all four controllers:

- **Kubernetes API server** for list/watch and the [`apiCall` context variable](/docs/policy-types/cluster-policy/external-data-sources#variables-from-kubernetes-api-server-calls)
- **DNS**
- **OCI registries** for [image verification](/docs/policy-types/cluster-policy/verify-images/overview) and [image registry context variables](/docs/policy-types/cluster-policy/external-data-sources#variables-from-image-registries)
- **External HTTPS endpoints** for [external service calls](/docs/policy-types/cluster-policy/external-data-sources#variables-from-service-calls) and the [CEL HTTP library](/docs/policy-types/cel-libraries#http-library). If you use plain HTTP for either, add a port 80 rule to the egress policy below.

> The samples use `app.kubernetes.io/instance: kyverno` and `app.kubernetes.io/part-of: kyverno` (the default Helm release name). If you used a different release name, replace both values in every manifest.

Reviewer: This statement says:
>If you used a different release name, replace both values in every manifest.

However, the chart assigns these labels differently:

* `app.kubernetes.io/instance` = Helm release name
* `app.kubernetes.io/part-of` = computed chart fullname

For example, `helm install prod kyverno/kyverno` normally produces:
```
app.kubernetes.io/instance: prod
app.kubernetes.io/part-of: prod-kyverno
```

Following the documentation and setting both labels to prod means none of the NetworkPolicies select the Kyverno pods.

Author: oh I see, nice catch, I updated the section to clarify further. also added `kubectl get pods --show-labels` if it helps

> The samples use `app.kubernetes.io/instance: kyverno` and `app.kubernetes.io/part-of: kyverno` (the default Helm release name). If you used a different release name, replace both values in every manifest.

Reviewer: ```suggestion
The samples use the labels from a default Kyverno Helm installation: `app.kubernetes.io/instance: kyverno` and `app.kubernetes.io/part-of: kyverno` (the default Helm release name). 
```

              kubernetes.io/metadata.name: monitoring
      ports:
        - protocol: TCP
          port: 8000
```

Apply the same manifest for the reports controller: change `metadata.name` to `kyverno-reports-controller` and change `app.kubernetes.io/component` in `spec.podSelector.matchLabels` to `reports-controller`.

##### Egress: all Kyverno controllers

One egress policy covers all four controllers. It allows DNS, the Kubernetes API server, and outbound HTTPS to public endpoints. It blocks the cloud metadata range (`169.254.0.0/16`) and RFC1918 private ranges (which typically include your pod and service CIDRs).

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kyverno-egress
  namespace: kyverno
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/part-of: kyverno
  policyTypes:
    - Egress
  egress:
    - to:
        - namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: kube-system
          podSelector:
            matchLabels:
              k8s-app: kube-dns
      ports:
        - protocol: UDP
          port: 53
        - protocol: TCP
          port: 53
    - to:
        - ipBlock:

>             cidr: 203.0.113.10/32

Reviewer: Kyverno requires Kubernetes API access for normal controller operation, not only when an `apiCall` context variable is used.

203.0.113.10/32 is a documentation placeholder from the RFC 5737 TEST-NET-3 range. Replace it with the CIDR or CIDRs your CNI evaluates for traffic from Kyverno Pods to the Kubernetes API server. Depending on the cluster and where NetworkPolicy is enforced relative to Service DNAT, this may be the `kubernetes.default` Service ClusterIP, the post-DNAT control-plane endpoint, or both. 

These commands can help identify the possible destinations:
```
# Kubernetes Service ClusterIP
kubectl get service kubernetes -n default \
  -o jsonpath='{.spec.clusterIP}{"\n"}'

# API server endpoints behind the Service
kubectl get endpoints kubernetes -n default -o wide
```

And replace that placeholder with one or more environment-specific ipBlock entries, and verify that all Kyverno controllers retain API connectivity after applying the policy.

Author: got it, I added those commands here
<img width="1169" height="485" alt="Screenshot 2026-07-16 at 9 45 35 PM" src="https://github.com/user-attachments/assets/f02befff-9841-4b57-95b7-e1113c6b41f3" />

- `monitoring` is a placeholder for the namespace hosting Prometheus. Replace it with your metrics-scrape namespace.
- The DNS rule assumes `k8s-app: kube-dns` pods in `kube-system`. This matches kubeadm, GKE, EKS, and AKS. On OpenShift, replace the podSelector with `dns.operator.openshift.io/daemonset-dns: default` and the namespaceSelector with `kubernetes.io/metadata.name: openshift-dns`.
- The `0.0.0.0/0` block excludes RFC1918 by default. RFC1918 typically covers your pod and service CIDRs. Add an explicit allow rule above it for in-cluster destinations such as a mirror registry or an [external service call](/docs/policy-types/cluster-policy/external-data-sources#variables-from-service-calls).

:::

##### Verification

Confirm the policies are applied and that admission still works. `--dry-run=server` exercises the full admission path without scheduling a pod, so it does not depend on PodSecurity or other cluster-level admission plugins:

```sh
kubectl get networkpolicy -n kyverno
kubectl run netpol-test --image=nginx --restart=Never --namespace=default --dry-run=server
```

> If the dry run fails with a webhook error (`failed calling webhook ... connection refused` or a timeout), the webhook ingress rule or the API server egress rule is likely mismatched. Check your CNI logs for policy-drop events, and verify the API server CIDR placeholder was replaced.

Reviewer: ```suggestion
Confirm that the NetworkPolicies are applied and that the API server can still reach the admission webhook:

```sh
kubectl get networkpolicy -n kyverno

kubectl run netpol-test \
  --image=nginx \
  --restart=Never \
  --namespace=default \
  --dry-run=server
```

The server-side dry run exercises the admission path without scheduling a Pod. It verifies admission-webhook reachability only; it does not verify Kyverno's outbound access to the Kubernetes API, DNS, registries, or external services.

After applying the egress policy, confirm that all Kyverno controller deployments remain available:

```sh
kubectl rollout status deployment --all -n kyverno
kubectl get pods -n kyverno
```

Inspect the controller logs for Kubernetes API connection failures, list/watch errors, DNS failures, or request timeouts:

```sh
kubectl logs -n kyverno \
  -l app.kubernetes.io/part-of=kyverno \
  --all-containers \
  --since=10m \
  --prefix
```

If image verification, image registry context variables, or external service calls are configured, test those features separately to verify that their required outbound destinations remain reachable.
```

Reviewer: Seems like there's a format issue, here's the paste-ready suggestion:

Confirm that the NetworkPolicies are applied and that the API server can still reach the admission webhook:

```sh
kubectl get networkpolicy -n kyverno

kubectl run netpol-test \
  --image=nginx \
  --restart=Never \
  --namespace=default \
  --dry-run=server
```

The server-side dry run exercises the admission path without scheduling a Pod. It verifies admission-webhook reachability only; it does not verify Kyverno's outbound access to the Kubernetes API, DNS, registries, or external services.

After applying the egress policy, confirm that all Kyverno controller deployments remain available:

```sh
kubectl rollout status deployment --all -n kyverno
kubectl get pods -n kyverno
```

Inspect the controller logs for Kubernetes API connection failures, list/watch errors, DNS failures, or request timeouts:

```sh
kubectl logs -n kyverno \
  -l app.kubernetes.io/part-of=kyverno \
  --all-containers \
  --since=10m \
  --prefix
```

If image verification, image registry context variables, or external service calls are configured, test those features separately to verify that their required outbound destinations remain reachable.

Author: gotcha, I made some update to clarify this section further with the commands ^

I confirmed all Kyverno controller deployments are available
note: i removed the `--all` from the rollout status deployment command, it doesn't seen to be a valid flag anymore
<img width="578" height="327" alt="Screenshot 2026-07-16 at 10 04 24 PM" src="https://github.com/user-attachments/assets/d3272c29-8482-4709-b8c7-a0539a3bb22e" />

> Inspect the controller logs for Kubernetes API connection failures, list/watch errors, DNS failures, or request timeouts

I did ran the command, didn't see any failures

<img width="1041" height="327" alt="Screenshot 2026-07-16 at 10 07 20 PM" src="https://github.com/user-attachments/assets/066dd594-f815-4ed1-bdb3-b51a47f579e8" />
